[circuits] add replacement data proof

s-ekai opened the pull request:
- In the input for Prove Data, you are required to enter the value for the data in the Replacement field.
- In Circom, it is set up to verify whether this data value matches the upper bits of the Replacement field stored as Provable Data.
- For testing purposes, I have created a function called genReplField to easily generate dummy data for the Replacement field.(If there are any unnatural aspects due to my unfamiliarity with TypeScript syntax, please let me know.)


fix: https://github.com/Unirep/create-unirep-app/issues/50

vivianjeng commented on packages/circuits/circuits/dataProof.circom:
```diff
@@ -2,10 +2,11 @@ pragma circom 2.1.0;
 
 
-template DataProof(STATE_TREE_DEPTH, FIELD_COUNT, SUM_FIELD_COUNT) {
+template DataProof(STATE_TREE_DEPTH, FIELD_COUNT, SUM_FIELD_COUNT, REPL_NONCE_BITS) {
```
Though it inherits unirep config
but in this template, we can still check 
```circom
assert(SUM_FIELD_COUNT < FIELD_COUNT);
```
to make sure `FIELD_COUNT - SUM_FIELD_COUNT` won't be overflowed

s-ekai commented:
The contract test was failing, so I will check it.

vivianjeng commented on packages/circuits/test/proveData.test.ts:
```diff
@@ -81,6 +82,18 @@ const genProofAndVerify = async (
     return { isValid, proof, publicSignals }
 }
 
+const genReplField = (value: number | bigint, length: number) => {
```
actually the repField could be `FIELD_COUNT-SUM_FIELD_COUNT` different data.
In the previous test (though might not be a good solution)
`Array(SUM_FIELD_COUNT).fill(6)`
 we know that it is an array fills with `6`

but in this function it might be confusing that we only insert one `value` into the array
I would suggest if we make it the function `genReplField`
we should let it be 
```ts
const genReplField = (values: (number | bigint)[]) => {}
```
and the length is `FIELD_COUNT - SUM_FIELD_COUNT` (could be fixed in the file)
then it can take the array input like `Array(SUM_FIELD_COUNT).fill(6)`

s-ekai commented:
@vivianjeng 

Thank you for the review!

- I've added conditions using assert.
- As for genReplField, I agree with your opinion, so I've modified it to take an array as an argument.
- I've also increased the number of publicSignals arguments on the contract side by the number of Replacement fields, as it was only accounting for the Addition fields.
- Additionally, I thought it would be better to test each item for both the Addition and Replacement fields, so I've expanded the test cases. (Though they are very simple tests and may not be the best approach.)

vivianjeng reviewed (approved):
Thank you so much. 🙏🏻
